third-party/utf8-decoder/utf8-decoder.h

```diff
@@ -1,12 +1,20 @@
+// FIXME: this check is probably too naive
+#if defined(__cplusplus)
+  #if __cplusplus >= 201103L
+    #define RESTRICT __restrict
+  #endif
+#else
+    #define RESTRICT restrict
+#endif
```

Comment:
AFAIK C++ has no `restrict` keyword in any form, and something like `__restrict` is compiler-specific.

Additionally your ifdefs leave RESTRICT undefined for C++ < 2011.

Additionally since this is a header file, we have to worry about symbol conflicts, so we should make it have a name like QIO_RESTRICT, e.g.

So I would recommend this:

``` c
#ifdef __cplusplus
  #define QIO_RESTRICT
#else
  #define QIO_RESTRICT restrict
#endif
```

I don't think we care so much about `restrict` in this code for the compiler so I think it's OK to leave it out for C++.